Port adaptive_avg_pool3d to ATen

crcrpar opened the pull request:
Resolves #18065.

crcrpar committed:
port Volumetric Adaptive Avg Pooling

crcrpar committed:
fix THNN/THCUNN header files

crcrpar committed:
Replace AveragePool with AdaptiveAveragePool

crcrpar committed:
Fix typos

crcrpar committed:
fix another typos

crcrpar committed:
fix typo

crcrpar committed:
fix end_index

crcrpar committed:
use AT_CUDA_CHECK & scalar_t

crcrpar commented on aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu:
```diff
@@ -0,0 +1,521 @@
+      });
+}
+
+void adaptive_avg_pool3d_backward_out_cuda_template(
```
@skrah @xmnlab 
Since this CUDA backward test fails and I'm struggling to fix this, could you take a look?

best regards

skrah commented on aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu:
```diff
@@ -0,0 +1,521 @@
+    Tensor& gradInput,
+    const Tensor& gradOutput_,
+    const Tensor& input) {
+  gradInput.resize_as_(input);
```
Resizing twice, also in the following function.

crcrpar replied:
ah oh, thank you!

skrah commented on aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu:
```diff
@@ -0,0 +1,521 @@
+
+  // somehow nonatomic is passing all test for volumetric case. (copied from
+  // THCUNN source code.)
+  // bool atomic = (isizeW%osizeW != 0) || (isizeH%osizeH != 0) || // (isizeT%osizeT != 0)
```
Can you try re-enabling line 448?

skrah replied:
I know you took it over from `THCUNN`, but it's enabled in `AdaptiveMaxPool3d` for instance.

crcrpar replied:
I tried and it worked. So I chose line 448.

skrah commented on aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu:
```diff
@@ -0,0 +1,521 @@
+  bool atomic = false;
+
+  if (input.ndimension() == 4) {
+    totalZ = sizeD * osizeT;
```
... and restoring these?

xmnlab commented on aten/src/ATen/native/AdaptiveAveragePooling3d.cpp:
```diff
@@ -0,0 +1,315 @@
+}
+
+void adaptive_avg_pool3d_out_cpu_template(
+    at::Tensor& output,
```
it is not a issue but just a comment, as it is also on the `at` namespace I think you can just use `Tensor` as you have use in other places in this file. but it is not something important .. just a comment

crcrpar replied:
Thank you. I like this kind of consistency and made changes.

xmnlab reviewed (commented):
as an overview it seems good. after you do the changes @skrah suggest, I will compare the old code with the new one to see if I find anything else.

skrah commented:
Just to be clear, I only tried to address the test failures. I haven't reviewed yet.

crcrpar committed:
use Tensor instead of at::Tensor

crcrpar committed:
Fix CUDA backward

Changes:
- Remove duplicate gradInput resizing
- Declare atomic based on the relation between isize* and osize*

crcrpar commented:
@skrah @xmnlab Thank you for your cooperation and now the problem seems resolved.

@ezyang could you take a look?

skrah commented:
@crcrpar Good that the issue is resolved!  I think we were asked to do a full review, so I'll start that (perhaps on Monday).  So far I've only focused on that one function. :)

skrah commented on aten/src/ATen/native/AdaptiveAveragePooling3d.cpp:
```diff
@@ -0,0 +1,314 @@
+  int osizeT = gradOutput_.size(-3);
+  int osizeH = gradOutput_.size(-2);
+  int osizeW = gradOutput_.size(-1);
+
```
The old code had `int64_t` here.  `AdaptiveAveragePooling.cpp` seems to use `int`, too, but in general I'd hesitate to change integer sizes during porting.

skrah commented on aten/src/ATen/native/AdaptiveAveragePooling3d.cpp:
```diff
@@ -0,0 +1,314 @@
+  int osizeW = gradOutput_.size(-1);
+
+  /* get contiguous gradOutput */
+  auto gradOutput = gradOutput_.contiguous();
```
Nitpick: The sizes should be the same, but I find it easier to use the sizes after conversion to contiguous, which is what the existing code did.

skrah commented on aten/src/ATen/native/AdaptiveAveragePooling3d.cpp:
```diff
@@ -0,0 +1,314 @@
+    int64_t b;
+#pragma omp parallel for private(b)
+    for (b = 0; b < input.size(0); b++) {
+      AT_DISPATCH_FLOATING_TYPES_AND_HALF(
```
The example PR for `AdaptiveAveragePooling` does it too, but actually dispatch should not happen inside the loop.

skrah commented on aten/src/ATen/native/AdaptiveAveragePooling3d.cpp:
```diff
@@ -0,0 +1,314 @@
+    Tensor& gradInput,
+    const Tensor& gradOutput_,
+    const Tensor& input) {
+  gradInput.resize_as_(input);
```
The old code also zeros `gradInput`.

skrah commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3657,15 +3657,25 @@
 
-- func: adaptive_avg_pool3d_backward(Tensor grad_output, Tensor self, *, Tensor(a!) grad_input) -> Tensor(a!)
-  python_module: nn
+- func: adaptive_avg_pool3d_backward(Tensor grad_output, Tensor self, *, Tensor(a!) gradInput) -> Tensor(a!)
```
It looks like the rest of the file uses `grad_input`.

skrah commented on aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu:
```diff
@@ -0,0 +1,517 @@
+  int64_t osizeT = output_size[0];
+  int64_t osizeH = output_size[1];
+  int64_t osizeW = output_size[2];
+
```
If you allow `output.size.size() == 1`, you also have to set `osizeH` and `osizeW` conditionally.

But I think the `JIT` issue does not happen for `AdaptiveAveragePooling`, so just checking for `output_size.size() == 3` may be sufficient if the tests pass.

ezyang replied:
Can we file an issue about this, and reference it to all of the cases where we added extra checks like this? I'm pretty sure this code is supposed to not be necessary, but since no one is root causing it, it looks like this code is going to replicate everywhere, can we at least put a comment on it.

skrah replied:
Yes, it's a good idea to file an issue, I'll open one.  Probably for this file the workaround is not necessary, but I haven't verified.

skrah commented:
@crcrpar I looks great to me in general, I've left some comments.

One completely general issue: Personally I'd find it easier to review if existing formatting is not changed during the porting step, for example `diff -uw existing.c new.cpp` is smaller.

But that is no change request, others PRs do it too.

crcrpar commented:
@skrah Thanks for the review.

I've reflected your comments. I moved dispatch outside the loop  following `AdaptiveMaxPooling` implementation.

> One completely general issue: Personally I'd find it easier to review if existing formatting is not changed during the porting step, for example `diff -uw existing.c new.cpp` is smaller.

Sorry to bother you by this. I applied clang-format partially in this PR.

skrah commented on aten/src/ATen/native/AdaptiveAveragePooling3d.cpp:
```diff
@@ -215,22 +250,46 @@ static void adaptive_avg_pool3d_backward_out_frame(
+  int64_t isizeW = input.size(-1);
+  int64_t osizeT = gradOutput_.size(-3);
+  int64_t osizeH = gradOutput_.size(-2);
+  int64_t osizeW = gradOutput_.size(-1);
```
Again, extreme nitpick, but it still calls `gradOutput_.size(-3)` and not `gradOutput.size(-3)`. :)

crcrpar replied:
Thanks.

skrah commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3657,15 +3657,25 @@
+    CUDA: adaptive_avg_pool3d_cuda
 
 - func: adaptive_avg_pool3d_backward(Tensor grad_output, Tensor self, *, Tensor(a!) grad_input) -> Tensor(a!)
-  python_module: nn
```
I'm curious, why did you remove `python_module: nn` here but not above? Apparently it builds just fine, but most other locations in this file still have `python_module: nn`.

crcrpar replied:
Oh, this is my fault. I did not remove this deliberately and I get this back for consistency with others.

skrah commented:
@crcrpar Thanks, I have added one additional question about `native_functions.yaml` and one small nitpick.

skrah commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3657,15 +3657,26 @@
+    CUDA: adaptive_avg_pool3d_backward_out_cuda
 
 - func: adaptive_avg_pool3d_backward(Tensor grad_output, Tensor self) -> Tensor
-  python_module: nn
```
This one is still missing.

crcrpar replied:
thanks... :bowing_man:

skrah commented:
Thanks, looks great!   Let's ask @ezyang to take over.

ezyang commented:
Thanks @skrah. As a methodological note, if you want to "accept" a review but pass it off to someone else for a complete review, instead of "accepting" it, just remove yourself from the reviewer list and add a comment LGTM. This prevents the PR as showing up as "accepted" in the UI, so that I know it does need review (in this case, I saved the email notification in my inbox for review) :)

ezyang commented on aten/src/ATen/native/AdaptiveAveragePooling3d.cpp:
```diff
@@ -0,0 +1,371 @@
+inline int end_index(int a, int b, int c) {
+  return (int)std::ceil((float)((a + 1) * c) / b);
+}
+
```
Did we lose this comment? Was it inaccurate?

```
/*
 * Description:
 *    This function adaptively average pools an input 5D tensor along dimensions
 *     2, 3 and 4.
 *
 *    gridDim.y blocks work together on a single 2D output plane specified by
 *    (blockIdx.x + offsetZ).
 */
```

crcrpar replied:
Hmm, I guess you got a bit confused while reviewing because I found only the below comment in THNN code (C impl).
https://github.com/pytorch/pytorch/blob/cf55670bddf6bb5f6199df2c543c8fb4ca21b08e/aten/src/THNN/generic/VolumetricAdaptiveAveragePooling.c#L10

Also, the new `.cu` file has the comment of description: https://github.com/pytorch/pytorch/pull/19898/files#diff-b9ff310d15f69d0b85cfe7fd87d8f8b9R33

ezyang commented on aten/src/ATen/native/AdaptiveAveragePooling3d.cpp:
```diff
@@ -0,0 +1,371 @@
+    int64_t istrideH,
+    int64_t istrideW) {
+  int64_t d;
+#pragma omp parallel for private(d)
```
This parallelization is new :) How much speed up did we get from it?

ezyang commented:
Everything looks in order, and review comments look addressable in follow up PRs.

crcrpar commented:
@ezyang Thanks for the look. Could you check comments (https://github.com/pytorch/pytorch/pull/19898#discussion_r281642657)?

ezyang commented:
You're right!  My bad.

Excerpts from Masaki Kozuki's message of 2019-05-07 09:18:30 -0700:
> crcrpar commented on this pull request.
> 
> > +#include <ATen/ATen.h>
> +#include <ATen/NativeFunctions.h>
> +
> +namespace at {
> +namespace native {
> +
> +namespace {
> +
> +inline int start_index(int a, int b, int c) {
> +  return (int)std::floor((float)(a * c) / b);
> +}
> +
> +inline int end_index(int a, int b, int c) {
> +  return (int)std::ceil((float)((a + 1) * c) / b);
> +}
> +
> 
> Hmm, I guess you got a bit confused while reviewing because I found only the below comment in THNN code (C impl).
> https://github.com/pytorch/pytorch/blob/cf55670bddf6bb5f6199df2c543c8fb4ca21b08e/aten/src/THNN/generic/VolumetricAdaptiveAveragePooling.c#L10
> 
> Also, the new `.cu` file has the comment of description: https://github.com/pytorch/pytorch/pull/19898/files#diff-b9ff310d15f69d0b85cfe7fd87d8f8b9R33
>

crcrpar committed:
Get sizes from contiguous gradOutput

ezyang commented:
@crcrpar I saw you pushed new commits (which caused the in progress land to fail). Were there any changes?

crcrpar commented:
@ezyang I found and fixed conflicts in `native_functions.yaml`.
